### ISSUE TITLE
requests.exceptions.TooManyRedirects: Exceeded 30 redirects... sessions.py's resolve_directs

### ISSUE THREAD

**Issue by antdude** [sopel-irc/sopel#1147]
...
04:26PM <Mousey> http://blackboxvoting.org/fraction-magic-1/ -- RIGGED!!
04:26PM <URL> requests.exceptions.TooManyRedirects: Exceeded 30 redirects. (file
          "/home/ant/SopelPython3/lib/python3.4/site-packages/requests/sessions.py", line 110, in
          resolve_redirects)
...
[05:03pm] <Ant> http://blackboxvoting.org/fraction-magic-1/
05:03PM <URL> requests.exceptions.TooManyRedirects: Exceeded 30 redirects. (file
          "/home/ant/SopelPython3/lib/python3.4/site-packages/requests/sessions.py", line 110, in
          resolve_redirects)
...

**Comment by dgw** [comment]
That particular server definitely sends redirects for certain user agents. Compare its behavior when presented with cURL's default UA string, and the Requests library default for the system version on the laptop I'm using:

````
$ python -i
Python 2.7.12 (default, Sep 28 2016, 18:38:03) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> print requests.utils.default_user_agent()
python-requests/2.10.0
>>> quit()

$ curl --head http://blackboxvoting.org/fraction-magic-1/
HTTP/1.1 200 OK
Date: Tue, 01 Nov 2016 00:42:44 GMT
Server: Apache/2.4.18 (Unix) OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.5.35
X-Powered-By: PHP/5.5.35
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Link: <http://blackboxvoting.org/wp-json/>; rel="https://api.w.org/", <http://blackboxvoting.org/?p=2914>; rel=shortlink
Set-Cookie: PHPSESSID=d6c6507adf1afe95320de91c19b41975; path=/
Vary: User-Agent
Content-Type: text/html; charset=UTF-8

$ curl --head -H "User-Agent: python-requests/2.10.0" http://blackboxvoting.org/fraction-magic-1/
HTTP/1.1 302 Found
Date: Tue, 01 Nov 2016 00:42:27 GMT
Server: Apache/2.4.18 (Unix) OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.5.35
Location: http://blackboxvoting.org/
Content-Type: text/html; charset=iso-8859-1
````

**Comment by antdude** [comment]
dgw -- FYI on mine:

$ python -i
Python 3.4.2 (default, Oct  8 2014, 10:45:20)
[GCC 4.9.1] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import requests
>>> print requests.utils.default_user_agent()
  File "<stdin>", line 1
    print requests.utils.default_user_agent()
                 ^
SyntaxError: invalid syntax
>>>
(SopelPython3)ant@ANTian:~/Downloads$ curl --head http://blackboxvoting.org/fraction-magic-1/
HTTP/1.1 200 OK
Date: Tue, 01 Nov 2016 00:55:37 GMT
Server: Apache/2.4.18 (Unix) OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.5.35
X-Powered-By: PHP/5.5.35
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Link: <http://blackboxvoting.org/wp-json/>; rel="https://api.w.org/", <http://blackboxvoting.org/?p=2914>; rel=shortlink
Set-Cookie: PHPSESSID=f08b5dee8290e2c9fbb3cc430e0126a3; path=/
Vary: User-Agent
Content-Type: text/html; charset=UTF-8

(SopelPython3)ant@ANTian:~/Downloads$ curl --head -H "User-Agent: python-requests/2.10.0" http://blackboxvoting.org/fraction-magic-1/
HTTP/1.1 302 Found
Date: Tue, 01 Nov 2016 00:55:42 GMT
Server: Apache/2.4.18 (Unix) OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.5.35
Location: http://blackboxvoting.org/
Content-Type: text/html; charset=iso-8859-1

**Comment by dgw** [comment]
The only material difference I see is that your default Python version is 3.4, in which `print` is a function (use `print(requests.utils.default_user_agent())` instead). The server is still redirecting when presented with the Requests user agent. I added the `-L` option to cURL (which instructs cURL to follow redirects) and it does in fact result in a redirect loop: `curl: (47) Maximum (50) redirects followed`

This doesn't seem to be a bug in Sopel, or even in Requests. It's an issue with the server's configuration redirecting certain user agents to the home page, even if they requested the home page.

As a workaround, Sopel could theoretically ask Requests to use a custom user agent when fetching pages, I suppose.

````
$ curl -s -L --head -H "User-Agent: Sopel/6.4.1" http://blackboxvoting.org/fraction-magic-1/
HTTP/1.1 200 OK
Date: Tue, 01 Nov 2016 14:01:30 GMT
Server: Apache/2.4.18 (Unix) OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.5.35
X-Powered-By: PHP/5.5.35
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Link: <http://blackboxvoting.org/wp-json/>; rel="https://api.w.org/", <http://blackboxvoting.org/?p=2914>; rel=shortlink
Set-Cookie: PHPSESSID=c7d5184fbd517ffa38b92541352c66cd; path=/
Vary: User-Agent
Content-Type: text/html; charset=UTF-8
````

**Comment by dasu** [comment]
I think a custom user agent is the way to go.  It would also resolve #1143

**Comment by embolalia** [comment]
Fixed in c8ef3d5
